File: easy/cpp/Question1.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;
/**
 * 方法一：双层遍历
 */
// class Solution {
//    public:
//     vector<int> twoSum(vector<int>& nums, int target) {
//         for (int i = 0; i < nums.size(); i++) {
//             for (int j = i + 1; j < nums.size(); j++) {
//                 if (nums[i] + nums[j] == target) {
//                     return {i, j};
//                 }
//             }
//         }
//         return {};
//     }
// };

/**
 * 方法二：枚举右维护左+哈希表
 */
class Solution {
   public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> map;
        for (int i = 0; i < nums.size(); i++) {
            if (map.find(target - nums[i]) != map.end()) {
                return {map[target - nums[i]], i};
            }
            map[nums[i]] = i;
        }
        return {};
    }
};

int main() {
    vector<int> nums = {2, 7, 11, 15};
    for (int x : Solution().twoSum(nums, 9)) {
        cout << x << " ";
    }
    cout << endl;
    return 0;
}

File: easy/java/Question1.java
package leetcode.easy;

import java.util.HashMap;
import java.util.Map;

import util.input.ArrayUtil;

public class Question1 {
    public static void main(String[] args) {
        class Solution {
            public int[] twoSum(int[] nums, int target) {
                Map<Integer, Integer> map = new HashMap<>();
                for (int i = 0; i < nums.length; i++) {
                    int expect = target - nums[i];
                    if (map.containsKey(expect)) {
                        return new int[] { map.get(expect), i };
                    }
                    map.put(nums[i], i);
                }
                return new int[0];
            }
        }
        ArrayUtil.showArray(new Solution().twoSum(new int[] { 2, 7, 11, 15 }, 9));
    }
}
